extensions/librdkafka/ConsumeKafka.cpp

```diff
@@ -0,0 +1,578 @@
+  rd_kafka_headers_t* headers_raw;
+  const rd_kafka_resp_err_t get_header_response = rd_kafka_message_headers(message, &headers_raw);
+  if (RD_KAFKA_RESP_ERR__NOENT == get_header_response) {
+    return {};
+  }
+  if (RD_KAFKA_RESP_ERR_NO_ERROR != get_header_response) {
+    logger_->log_error("Failed to fetch message headers: %d: %s", rd_kafka_last_error(), rd_kafka_err2str(rd_kafka_last_error()));
+  }
+  std::vector<std::string> matching_headers;
+  for (std::size_t header_idx = 0;; ++header_idx) {
+    const char* value;  // Not to be freed
+    std::size_t size;
+    if (RD_KAFKA_RESP_ERR_NO_ERROR != rd_kafka_header_get(headers_raw, header_idx, header_name.c_str(), (const void**)(&value), &size)) {
+      break;
+    }
+    if (size < 200) {
+      logger_->log_debug("%.*s", static_cast<int>(size), value);
+    } else {
+      logger_->log_debug("%.*s...", 200, value);
+    }
+    matching_headers.emplace_back(value, size);
+  }
+  return matching_headers;
+}
+
+std::vector<std::pair<std::string, std::string>> ConsumeKafka::get_flowfile_attributes_from_message_header(const rd_kafka_message_t* message) const {
+  std::vector<std::pair<std::string, std::string>> attributes_from_headers;
+  for (const std::string& header_name : headers_to_add_as_attributes_) {
+    const std::vector<std::string> matching_headers = get_matching_headers(message, header_name);
+    if (matching_headers.size()) {
+      attributes_from_headers.emplace_back(header_name, utils::get_encoded_string(resolve_duplicate_headers(matching_headers), message_header_encoding_attr_to_enum()));
+    }
+  }
+  return attributes_from_headers;
+}
+
+class WriteCallback : public OutputStreamCallback {
+ public:
+  WriteCallback(char *data, uint64_t size) :
+      data_(reinterpret_cast<uint8_t*>(data)),
+      dataSize_(size) {}
+  uint8_t* data_;
+  uint64_t dataSize_;
+  int64_t process(const std::shared_ptr<io::BaseStream>& stream) {
+    int64_t ret = 0;
+    if (data_) {
+      ret = stream->write(data_,  gsl::narrow<int>(dataSize_));
+    }
+    return ret;
+  }
+};
+
+void ConsumeKafka::add_kafka_attributes_to_flowfile(std::shared_ptr<FlowFileRecord>& flow_file, const rd_kafka_message_t* message) const {
+  // We do not currently support batching messages into a single flowfile
+  flow_file->setAttribute(KAFKA_COUNT_ATTR, "1");
+  const utils::optional<std::string> message_key = utils::get_encoded_message_key(message, key_attr_encoding_attr_to_enum());
+  if (message_key) {
+    flow_file->setAttribute(KAFKA_MESSAGE_KEY_ATTR, message_key.value());
+  }
+  flow_file->setAttribute(KAFKA_OFFSET_ATTR, std::to_string(message->offset));
+  flow_file->setAttribute(KAFKA_PARTITION_ATTR, std::to_string(message->partition));
+  flow_file->setAttribute(KAFKA_TOPIC_ATTR, rd_kafka_topic_name(message->rkt));
+}
+
+utils::optional<std::vector<std::shared_ptr<FlowFileRecord>>> ConsumeKafka::transform_pending_messages_into_flowfiles(core::ProcessSession* session) const {
```

Comment:
This should probably take a reference instead of a pointer, to avoid null pointer problems. Please check all functions taking a raw pointer, except for onTrigger/onSchedule, which are constrained by the base class interface.

In general, we should make sure that whenever we're working with a nullable pointer, we don't dereference it without checking in the same scope. When we don't need nullability, references (for passing) or gsl::not_null (for storage) can be used instead.

Reply:
My reasoning on passing pointers for these functions was that this is how the processor is get called as well, so this was the simplest way to forward the args. I assumed that by the time the onTrigger is called on the processors in a flow the session is already checked for validity.

Reply:
I thought so, but it improves readability and therefore safety when you can verify that locally.

Reply:
Updated the helper functions to take session by reference.